### ISSUE TITLE
ignore specified modules when running xxx-all commands (plan-all, apply-all, etc.)

### ISSUE THREAD

**Issue by ebarault** [gruntwork-io/terragrunt#355]
Sometime it might be handy to prevent some modules from being handled by terragrunt **xxx-all** commands in a terragrunt project hierarchy while keeping the project hierarchy as-is for the sake of keeping the remote states coherent.

Consider this project structure:
```
root
├── app
│   ├── main.tf
│   └── terraform.tfvars
├── db
│   ├── main.tf
│   └── terraform.tfvars
└── vpc
    ├── main.tf
    └── terraform.tfvars
```

Say for example when i execute `terragrunt apply-all` at the root I want **app** and **db** to be handled, but not **vpc** which I manage separately. I think this is not currently supported.

Maybe a new directive inside the terragrunt config block could be used to indicate a given module should be ignored by `xxx-all` commands?

```
terragrunt = {
  dependencies {
    ignore = true
  }
}
```

**Comment by brikis98** [comment]
I've run into similar circumstances before. Putting some in the `.tfvars` file is one option. Another is a CLI flag along the lines of `--terragrunt-ignore some/path/to/ignore`.

**Comment by ebarault** [comment]
yes, i thought about this option too. that would fit too and be probably more flexible as it would enable multiple scenarios with the same terragrunt repo, which would be nice for CI/CD integration :
- job **A** loads the full repo and triggers infrastructure components deploy only
- job **B** loads the same repo and triggers service components deploy only

so +1 for `--terragrunt-ignore`

**Comment by brikis98** [comment]
I'd definitely welcome a PR for that :)

**Comment by ebarault** [comment]
I don't know go, but I can try if you considerer it feasible (entry-level PR) and provide a starting point

**Comment by brikis98** [comment]
It's probably closer to intermediate rather than entry-level. But key points are:

* [Argument parsing](https://github.com/gruntwork-io/terragrunt/blob/master/cli/args.go#L18).
* [Where modules are found for the xxx-all command](https://github.com/gruntwork-io/terragrunt/blob/master/configstack/module.go#L38). 

You'd read store the `--terragrunt-ignore` arguments in `TerragruntOptions` and then in `ResolveTerraformModules`, skip over any modules that match the ignore args.

**Comment by ebarault** [comment]
ok, thanks, I'll try to have a look in the coming days

**Comment by tomdavidson** [comment]
What about globs for the workdir instead? For the OP it could be:

`--terragrunt-working-dir !(vpc)` or `--terragrunt-working-dir !(app|db)`

**Comment by brikis98** [comment]
Yea, I guess supporting globs in a `--terragrunt-exclude` parameter would be the most "unix" friendly way of doing it.

**Comment by tomdavidson** [comment]
GoLang has globs and matching that is similar but not the same as git/nodejs/unix

- https://golang.org/pkg/path/filepath/#Glob
- http://pubs.opengroup.org/onlinepubs/7908799/xsh/glob.html

Maybe the consistency would be worth a dependency such as (keep in mind, I have never done golang):

- https://github.com/gobwas/glob

What is the advantage of an additional `--terragrunt-exclude` flag vs globbing `--terragrunt-working-dir`?

**Comment by brikis98** [comment]
> GoLang has globs and matching that is similar but not the same as git/nodejs/unix

We already use https://github.com/mattn/go-zglob. Could switch to something else. That said, wouldn't the glob be expanded by the OS rather than Terragrunt?

> What is the advantage of an additional --terragrunt-exclude flag vs globbing --terragrunt-working-dir?

For the `xxx-all` commands (e.g., `apply-all`), `--terragrunt-working-dir` currently just sets where to search for submodules. Are you proposing changing that behavior to instead explicitly specify (via globs) which modules to use? I guess in that case, the default glob is something like `**/terraform.tfvars`... 

I'm not against it. The reality is we eventually may need both. `--terragrunt-working-dir` specifies what to include and `--terragrunt-exclude` specifies what to exclude.

> The glob would survive parseStringArg

**Comment by tomdavidson** [comment]
Oh man, I might not have a clue - the OS expands the glob? I thought the glob was just a string argument passed into the lib which had to sort it out by filtering.

go-zglob supports ** and I'm not suggesting a switch.

Reducing input flags is not this issue. I think the better point would be be first class globbing could have impact beyond --terragrunt-exclude (the exclude flag would work just fine for me). Terragrunt would need to translate a found tfvars file to a dirname - doesn't tg already do this?  In bash I would do something like:
 `find -mindepth 2 -not -path "*.terraform*" -name '$GLOB*.tfvars' -printf '%P\n' | xargs -I % dirname %`

The example raised a question. What about matching on the file name such as  (**/aws/*/prod.tfvars) or on dirname such as (**/prod-*) ? If there was not an .ext at the end of the glob, maybe TG appends *.tfvars to it?

**Comment by brikis98** [comment]
> Oh man, I might not have a clue - the OS expands the glob? I thought the glob was just a string argument passed into the lib which had to sort it out by filtering.

Generally, if you type something like:

```bash
some-command *.txt
```

Your shell will automatically expand the glob, in this case turning `*.txt` into the list of files with the extension `.txt` in the current folder. So, we could just rely on the OS for this functionality. Otherwise, you'd have to pass any globs to Terragrunt wrapped in single quotes (`*.tfvars`) so the shell *doesn't* expand them.

> Terragrunt would need to translate a found tfvars file to a dirname - doesn't tg already do this?

Yes, it does. I'm merely pointing out it would be weird for a parameter called `--working-dir` (emphasis on "dir") to use a glob like `**/terraform.tfvars`, as that doesn't point to a directory, but to files. And yet without the `terraform.tfvars`, there's no obvious way to find "all" modules.

**Comment by tomdavidson** [comment]
Ahh I understand what you are saying now. The globbed input would need to be one param. This make sense and I think is typical.

 > Yes, it does. I'm merely pointing out it would be weird for a parameter called --working-dir (emphasis on "dir") to use a glob like **/terraform.tfvars, as that doesn't point to a directory, but to files. And yet without the terraform.tfvars, there's no obvious way to find "all" modules.

Use the terraform.tfvars file to file the dirs - just like the bash one liner example. I assumed TG already does this. I think following bash will give a list of all directories that contain a .tfvar matching glob with a default of '**/terraform.tfvars' but that takes globs of dirnames (just to illustrate):

``` sh
function find-dirs () {
  local glob=${1:-'**/terrafrom.tfvars}
  local dirs
  if [[ ! glob =~ (.tfvars)$ ]]; then glob=$glob/terraform.tfvars; fi
  find -mindepth 2 -not -path "*.terraform*" -name '$glob' -printf '%P\n' | xargs -I % dirname %P 
}

find-dirs "**/(app|db)"
```

**Comment by brikis98** [comment]
Yea, you can definitely do fancy bash code to pull out the dirnames. The key questions are (a) what is the most common use case here and (b) what is the simplest user experience we could create for that use case. 

If you were running `terragrunt apply-all` and wanted to customize which folders actually got applied, would the common pattern be to say:

1. Apply to these specific folders
1. Apply to all modules, but exclude these specific folders

While both items are useful, my *guess* is the second item will be more common. That makes me think a `--terragrunt-exclude` param could be a good starting point. The `--working-dir` tweak could be a follow-on.

The next question is how to specify what to include/exclude. The options are:

1. Glob expanded by the shell.
1. Glob expanded by Terragrunt.
1. Regex matched by Terragrunt.
1. `.gitignore` syntax matched by Terragrunt. Similar to glob, but not exactly.

I suspect for excluding things, `.gitignore` is probably the most convenient syntax, but I'm open to opinions. 

@ebarault and @tomdavidson If you have concrete use cases to share, perhaps that would help us figure out the most efficient way to make them work.

**Comment by tomdavidson** [comment]
The OP has a clear concrete case which could be solved with globs such as:

`--terragrunt-working-dir !(vpc)` or `--terragrunt-working-dir !(app|db)`


I can not just `validate-all  configs/ ` in the status quo because the DRY terragrunt config has:

```hcl
  terraform {
    extra_arguments "common_vars" {
      commands = ["${get_terraform_commands_that_need_vars()}"]

      required_var_files = [
        "${get_tfvars_dir()}/${find_in_parent_folders("common.tfvars", "skip-if-common.tfvars-not-exsit")}"
      ]
    }
  }
```

```
config
├── common.tfvars
├── pq
│   └── terraform.tfvars
├── prod
│   ├── hactar
│   │   ├── main.tf
│   │   └── terraform.tfvars
│   └── roosta
│       ├── main.tf
│       └── terraform.tfvars
├── stage
│   ├── aws
│   │   └── terraform.tfvars
│   ├── stage.tfvars
│   └── vmw
│       └── terraform.tfvars
└── terraform.tfvars
```

Instead I need to:
```
terragrunt validate-all $TG_CLI_ARGS config/pg
terragrunt validate-all $TG_CLI_ARGS config/stage/aws
terragrunt validate-all $TG_CLI_ARGS config/stage/vmw
terragrunt validate-all $TG_CLI_ARGS config/prod
```

Wouldn't `validate  configs/**/*` be slick?

**Comment by brikis98** [comment]
> The OP has a clear concrete case which could be solved with globs such as:

Obviously, I can envision cases where you need to include or exclude, but I want to hear real-world examples where not having this ability was a problem. It's not obvious to me if the VPC mentioned in the original post was an example or actual use-case. If it's the latter, then great, having a way to exclude a folders from `xxx-all` commands seems like the solution we need. I'd love to see more real-world use cases so we can see if that generalizes.

> I can not just validate-all configs/ in the status quo because the DRY terragrunt config has:

I'm still not clear on why this is failing for you?

**Comment by ebarault** [comment]
hi, sorry for radio silence.
Actually my original UC would be more of the **ignore** type.
I'm building self contained projects using several service components + 1 codebuild/codepipeline project.
My initial intention is to perform apply on all components but the CI one which i update manually, or based on a flag in the CI scripts.

Considering that I can have a arbitrary number of service components inside a project, the idea is more the exclude the CI component rather than including each service components. 

But of course an `--include-dir !(ci)` option would do just as well, but seems much harder to me to implement.

**Comment by gfodor** [comment]
Hey just chiming in here, that I agree it would be useful to exclude modules entirely from an environment when building up/tearing down things with terragrunt. The example we are hitting (ironically) is we want our CI subsystems to only exist in our dev environment, and not our prod environment. So a simple blacklist of modules would work great.

**Comment by brikis98** [comment]
OK, then some sort of exclude functionality seems to be like a great first step.

Anyone up for a PR?

**Comment by stonefury** [comment]
My case I want to deploy the networking separately, and would just prefer to not have it included as part of apply-all.

**Comment by msvechla** [comment]
@brikis98 I just added a PR which implements exactly this functionality, see: https://github.com/gruntwork-io/terragrunt/pull/563

Please let me know what you think and what needs to be done to get this merged.

**Comment by brikis98** [comment]
Thank you @msvechla! I just left a bunch of comments in the PR.

**Comment by brikis98** [comment]
Fixed in #563. Thanks @msvechla!
